types/react-table/index.d.ts

```diff
@@ -1,746 +1,673 @@
+    extends Omit<TableOptions<D>, 'columns' | 'pageCount'>,
+        UseTableInstanceProps<D> {}
+
+export interface TableState<D extends object = {}> {
+    hiddenColumns?: Array<IdType<D>>;
 }
-export interface Filter {
-    id: string;
-    value: any;
-    pivotId?: string;
+
+export interface Hooks<D extends object = {}> extends UseTableHooks<D> {}
+
+export interface Cell<D extends object = {}> extends UseTableCellProps<D> {}
+
+export interface Column<D extends object = {}> extends UseTableColumnOptions<D> {}
+
+export interface ColumnInstance<D extends object = {}>
+    extends Omit<Column<D>, 'id' | 'columns'>,
+        UseTableColumnProps<D> {}
+
+export interface HeaderGroup<D extends object = {}> extends ColumnInstance<D>, UseTableHeaderGroupProps<D> {}
+
+export interface Row<D extends object = {}> extends UseTableRowProps<D> {}
+
+/* #region useTable */
+export function useTable<D extends object = {}>(
+    options: TableOptions<D>,
+    ...plugins: Array<PluginHook<D>>
+): TableInstance<D>;
+
+/**
+ * NOTE: To use custom options, use "Interface Merging" to add the custom options
+ */
+export type UseTableOptions<D extends object> = {
+    columns: Array<Column<D>>;
+    data: D[];
+} & Partial<{
+    initialState: Partial<TableState<D>>;
+    reducer: (newState: TableState<D>, action: string, previousState: TableState<D>) => TableState<D>;
+    useControlledState: (state: TableState<D>) => TableState<D>;
+    defaultColumn: Partial<Column<D>>;
+    initialRowStateKey: IdType<D>;
+    getSubRows: (originalRow: D, relativeIndex: number) => D[];
+    getRowId: (originalRow: D, relativeIndex: number) => IdType<D>;
+}>;
+
+export interface UseTableHooks<D extends object> {
+    useOptions: Array<(options: TableOptions<D>, args: TableOptions<D>) => TableOptions<D>>;
+    stateReducers: Array<
+        (
+            newState: TableState<D>,
+            action: string,
+            previousState: TableState<D>,
+            instance: RefObject<TableInstance<D>>,
+        ) => TableState<D>
+    >;
+    columns: Array<(columns: Array<Column<D>>, instance: TableInstance<D>) => Array<Column<D>>>;
+    columnsDeps: Array<(deps: any[], instance: TableInstance<D>) => any[]>;
+    flatColumns: Array<(flatColumns: Array<Column<D>>, instance: TableInstance<D>) => Array<Column<D>>>;
+    flatColumnsDeps: Array<(deps: any[], instance: TableInstance<D>) => any[]>;
+    headerGroups: Array<(flatColumns: Array<HeaderGroup<D>>, instance: TableInstance<D>) => Array<HeaderGroup<D>>>;
+    headerGroupDeps: Array<(deps: any[], instance: TableInstance<D>) => any[]>;
+    useInstanceBeforeDimensions: Array<(instance: TableInstance<D>) => void>;
+    useInstance: Array<(instance: TableInstance<D>) => void>;
+    useRows: Array<(rows: Array<Row<D>>, instance: TableInstance<D>) => Array<Row<D>>>;
+    prepareRow: Array<(row: Row<D>, instance: TableInstance<D>) => Row<D>>;
+    getTableProps: Array<(props: object, instance: TableInstance<D>) => object>;
```

Comment:
This and other prop getters below now can be also just objects or arrays of objects instead of callbacks, and the callbacks can also return arrays of objects, see more there: https://github.com/tannerlinsley/react-table/issues/1749#issuecomment-564177091

Reply:
yeah, I just noticed this.

Reply:
I refactored the types for all of the property getters, there might be a more compact way to have done it, but what's there is at least correct, I hope :)

Happy to take any suggestions about how to make them cleaner.

Reply:
Here is my PR for your PR :) https://github.com/ggascoigne/DefinitelyTyped/pull/1 — adds a generic that makes it much easier to create all the getters (and also fixes their signatures etc.)